### PR TITLE
[JavaScript] Improve exercise: concept/arrays #1300

### PR DESCRIPTION
**Pull request by peterchu999**
i've changing the exercise based on the issue approach, But not yet implemented the exercise for transforms / loops and analysis. #1300

### CODE REVIEW

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 1]
```diff
@@ -1,11 +1,13 @@
...
   setItem,
-  prefilledArray,
+  insertItemFromTop,
   removeItem,
   removeItemFromTop,
+  insertItemFromBottom,
+  removeItemFromBottom,
   checkLengthOfStack
-} from './arrays';
+} from './example';
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:10]:
```suggestion
} from './arrays';
```

File: languages/javascript/exercises/concept/arrays/example.js [hunk 2]
```diff
@@ -30,6 +22,16 @@ export function removeItemFromTop(array) {
   return array;
 }
 
+export function insertItemFromBottom(array, newCard) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/example.js:25]:
```suggestion
export function insertItemAtBottom(array, newCard) {
```

File: languages/javascript/exercises/concept/arrays/example.js [hunk 3]
```diff
@@ -7,17 +7,9 @@ export function setItem(array, index, newValue) {
...
-    return [];
-  }
-
-  const prefilledArray = [];
-  for (let i = 0; i < length; i++) {
-    prefilledArray.push(value);
-  }
-
-  return prefilledArray;
+export function insertItemFromTop(array, newCard) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/example.js:10]:
```suggestion
export function insertItemAtTop(array, newCard) {
```

File: languages/javascript/exercises/concept/arrays/suggestion.arrays.spec.js [hunk 4]
```diff
@@ -0,0 +1,140 @@
+import {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/suggestion.arrays.spec.js:1]: What is this file :) ?

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 5]
```diff
@@ -80,6 +82,36 @@ describe('arrays', () => {
...
 
+  describe('insertItemFromBottom', () => {
+    const insertItemFromBottomCases = [
+      [[1], 2, [2, 1]],
+      [[3, 1, 2], 1, [1, 3, 1, 2]],
+      [[9, 9, 9, 9], 9, [9, 9, 9, 9, 9]],
+      [[5, 9, 7, 1], 8, [8, 5, 9, 7, 1]]
+    ];
+
+    insertItemFromBottomCases.forEach(([array, newCard, expected]) => {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:93]:
```suggestion
    insertItemAtBottomCases.forEach(([array, newCard, expected]) => {
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 6]
```diff
@@ -80,6 +82,36 @@ describe('arrays', () => {
     });
   });
 
+  describe('insertItemFromBottom', () => {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:85]:
```suggestion
  describe('insertItemAtBottom', () => {
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 7]
```diff
@@ -80,6 +82,36 @@ describe('arrays', () => {
...
+    const insertItemFromBottomCases = [
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:86]:
```suggestion
    const insertItemAtBottomCases = [
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 8]
```diff
@@ -80,6 +82,36 @@ describe('arrays', () => {
...
+      [[1], 2, [2, 1]],
+      [[3, 1, 2], 1, [1, 3, 1, 2]],
+      [[9, 9, 9, 9], 9, [9, 9, 9, 9, 9]],
+      [[5, 9, 7, 1], 8, [8, 5, 9, 7, 1]]
+    ];
+
+    insertItemFromBottomCases.forEach(([array, newCard, expected]) => {
+      test(`insertItemFromBottom([${array}], ${newCard})`, () => {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:94]:
```suggestion
      test(`insertItemAtBottom([${array}], ${newCard})`, () => {
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 9]
```diff
@@ -80,6 +82,36 @@ describe('arrays', () => {
...
+        expect(insertItemFromBottom(array, newCard)).toStrictEqual(expected);
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:95]:
```suggestion
        expect(insertItemAtBottom(array, newCard)).toStrictEqual(expected);
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 10]
```diff
@@ -29,24 +31,24 @@ describe('arrays', () => {
...
+      [[1], 2, [1, 2]],
+      [[2, 5], 3, [2, 5, 3]],
+      [[3, 4, 9, 7], 8, [3, 4, 9, 7, 8]],
+      [[5, 9, 7, 1], 8, [5, 9, 7, 1, 8]]
     ];
 
-    prefilledArrayTestCases.forEach(([value, length, expected]) => {
-      test(`prefilledArray([${value}], ${length})`, () => {
-        expect(prefilledArray(value, length)).toStrictEqual(expected);
+    insertItemFromTopCases.forEach(([array, newCard, expected]) => {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:49]:
```suggestion
    insertItemAtTopCases.forEach(([array, newCard, expected]) => {
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 11]
```diff
@@ -29,24 +31,24 @@ describe('arrays', () => {
...
+      test(`prefilledArray([${array}], ${newCard})`, () => {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:50]:
```suggestion
      test(`insertItemAtTop([${array}], ${newCard})`, () => {
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 12]
```diff
@@ -29,24 +31,24 @@ describe('arrays', () => {
...
+        expect(insertItemFromTop(array, newCard)).toStrictEqual(expected);
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:51]:
```suggestion
        expect(insertItemAtTop(array, newCard)).toStrictEqual(expected);
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 13]
```diff
@@ -29,24 +31,24 @@ describe('arrays', () => {
...
     });
   });
 
-  describe('prefilledArray', () => {
-    const prefilledArrayTestCases = [
-      [1, 0, []],
-      [2, 1, [2]],
-      [3, 2, [3, 3]],
-      [5, 6, [5, 5, 5, 5, 5, 5]]
+  describe('insertItemFromTop', () => {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:41]:
```suggestion
  describe('insertItemAtTop', () => {
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 14]
```diff
@@ -29,24 +31,24 @@ describe('arrays', () => {
...
+    const insertItemFromTopCases = [
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:42]:
```suggestion
    const insertItemAtTopCases = [
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 15]
```diff
@@ -1,11 +1,13 @@
 import {
   getItem,
   setItem,
-  prefilledArray,
+  insertItemFromTop,
   removeItem,
   removeItemFromTop,
+  insertItemFromBottom,
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:7]:
```suggestion
  insertItemAtBottom,
```

File: languages/javascript/exercises/concept/arrays/arrays.spec.js [hunk 16]
```diff
@@ -1,11 +1,13 @@
 import {
   getItem,
   setItem,
-  prefilledArray,
+  insertItemFromTop,
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.spec.js:4]:
```suggestion
  insertItemAtTop,
```

File: languages/javascript/exercises/concept/arrays/arrays.js [hunk 17]
```diff
@@ -1,23 +1,79 @@
...
+export function insertItemFromBottom(array, newCard) {
+  throw new Error('Implement the insertItemFromBottom function');
+}
+
+/**
+ * @param {number[]} array
+ * 
+ * @returns {number[]}
+ */
+export function removeItemFromBottom(array) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.js:66]:
```suggestion
export function removeItemAtBottom(array) {
```

File: languages/javascript/exercises/concept/arrays/arrays.js [hunk 18]
```diff
@@ -1,23 +1,79 @@
...
+  throw new Error('Implement the removeItemFromBottom function');
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.js:67]:
```suggestion
  throw new Error('Implement the removeItemAtBottom function');
```

File: languages/javascript/exercises/concept/arrays/arrays.js [hunk 19]
```diff
@@ -1,23 +1,79 @@
...
 }
 
-export function checkLengthOfStack(array, stackLength) {
+/**
+ * @param {number[]} array
+ * @param {number} newCard
+ * 
+ * @returns {number[]}
+ */
+export function insertItemFromBottom(array, newCard) {
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.js:57]:
```suggestion
export function insertItemAtBottom(array, newCard) {
```

File: languages/javascript/exercises/concept/arrays/arrays.js [hunk 20]
```diff
@@ -1,23 +1,79 @@
...
+  throw new Error('Implement the insertItemFromBottom function');
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.js:58]:
```suggestion
  throw new Error('Implement the insertItemAtBottom function');
```

File: languages/javascript/exercises/concept/arrays/arrays.js [hunk 21]
```diff
@@ -1,23 +1,79 @@
+/**
+ * @param {number[]} array
+ * @param {number} index
+ * 
+ * @returns {number[]}
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/arrays.js:5]:
```suggestion
 * @returns {number} the item at position `index`  in the given `array`
```

Comment (SleeplessByte) [reply languages/javascript/exercises/concept/arrays/arrays.js:5]: Alternatively:

```suggestion
 * @returns {number}
```

And add a short description (1 line) to each function, at the top.

File: languages/javascript/exercises/concept/arrays/.docs/instructions.md [hunk 22]
```diff
@@ -54,7 +53,26 @@ removeItemFromTop([3, 2, 6, 4, 8])
...
+
+Insert new card at the bottom of the stack and return the stack.
+
+```javascript
+const newCard = 8
+insertItemFromBottom([5, 9, 7, 1], newCard)
+// => [8, 5, 9, 7, 1]
+```
+
+### 7. Remove a card in the bottom of the stack
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/.docs/instructions.md:66]:
```suggestion
### 7. Remove a card from the bottom of the stack
```

File: languages/javascript/exercises/concept/arrays/.docs/instructions.md [hunk 23]
```diff
@@ -54,7 +53,26 @@ removeItemFromTop([3, 2, 6, 4, 8])
 // => [3, 2, 6, 4]
 ```
 
-### 6. Check size of the stack
+### 6. Insert a card in the bottom of the stack
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/.docs/instructions.md:56]:
```suggestion
### 6. Insert a card at the bottom of the stack
```

File: languages/javascript/exercises/concept/arrays/.docs/instructions.md [hunk 24]
```diff
@@ -54,7 +53,26 @@ removeItemFromTop([3, 2, 6, 4, 8])
...
+
+Insert new card at the bottom of the stack and return the stack.
+
+```javascript
+const newCard = 8
+insertItemFromBottom([5, 9, 7, 1], newCard)
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/.docs/instructions.md:62]:
```suggestion
insertItemAtBottom([5, 9, 7, 1], newCard)
```

File: languages/javascript/exercises/concept/arrays/.docs/instructions.md [hunk 25]
```diff
@@ -24,15 +24,14 @@ setItem([1, 2, 4, 1], index, newCard)
...
-Create a stack of given `length` and fill it with cards of the given `value`.
+Insert new card at the top of the stack and return the stack.
 
 ```javascript
-const value = 8
-const length = 3
-prefilledSlice(value, length)
-// => [8, 8, 8]
+const newCard = 8
+insertItemFromTop([5, 9, 7, 1], newCard)
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/.docs/instructions.md:33]:
```suggestion
insertItemAtTop([5, 9, 7, 1], newCard)
```

File: languages/javascript/exercises/concept/arrays/.docs/instructions.md [hunk 26]
```diff
@@ -24,15 +24,14 @@ setItem([1, 2, 4, 1], index, newCard)
 // => [1, 2, 6, 1]
 ```
 
-### 3. Create a stack of cards
+### 3. Insert a card in the of top stack
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/.docs/instructions.md:27]:
```suggestion
### 3. Insert a card at the of top stack
```

File: languages/javascript/exercises/concept/arrays/suggestion.arrays.spec.js (hunk 4 above)
Comment (peterchu999) [reply languages/javascript/exercises/concept/arrays/suggestion.arrays.spec.js:1]: actually this is suggestion of the testcase that i make. what i'm trying to make here is to check that the student mutate the array not creating the new array @SleeplessByte

File: languages/javascript/exercises/concept/arrays/.docs/hints.md [hunk 27]
```diff
@@ -8,7 +8,7 @@
 - The array is a mutable structure, you can change its content anytime.
 - You can find an example [here][change_array_elements_resource], inside the 'Changing an Array Element' section.
 
-### 3. Create a stack of cards
+### 3. Insert a card at the of top stack
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/.docs/hints.md:11]:
```suggestion
### 3. Insert a card at the top of the stack
```

File: languages/javascript/exercises/concept/arrays/.docs/instructions.md [hunk 28]
```diff
@@ -24,15 +24,14 @@ setItem([1, 2, 4, 1], index, newCard)
 // => [1, 2, 6, 1]
 ```
 
-### 3. Create a stack of cards
+### 3. Insert a card at the of top stack
```
Comment (SleeplessByte) [inline languages/javascript/exercises/concept/arrays/.docs/instructions.md:27]:
```suggestion
### 3. Insert a card at the top of the stack
```

File: languages/javascript/exercises/concept/arrays/suggestion.arrays.spec.js (hunk 4 above)
Comment (SleeplessByte) [reply languages/javascript/exercises/concept/arrays/suggestion.arrays.spec.js:1]: You don't need to test for that.  Immutability / Mutability will be handled in a different exercise :)

Comment (peterchu999) [reply languages/javascript/exercises/concept/arrays/suggestion.arrays.spec.js:1]: okay then let just delete this file : )

File: languages/javascript/exercises/concept/arrays/arrays.js (hunk 17 above)
Comment (peterchu999) [reply languages/javascript/exercises/concept/arrays/arrays.js:80]: do i need to change the instruction.md function name too ?

**Comment by SleeplessByte** [comment]
Thank you very much @peterchu999 .

**Pull request merged**